File: backend/app/interfaces/schemas/usage.py
"""API schemas for usage tracking endpoints."""

from datetime import date, datetime

from pydantic import BaseModel


class TodayUsageResponse(BaseModel):
    """Today's usage summary."""

    tokens: int
    cost: float
    llm_calls: int
    tool_calls: int


class MonthUsageResponse(BaseModel):
    """This month's usage summary."""

    tokens: int
    cost: float
    llm_calls: int
    tool_calls: int
    sessions: int
    active_days: int


class UsageSummaryResponse(BaseModel):
    """Combined usage summary for today and this month."""

    today: TodayUsageResponse
    month: MonthUsageResponse


class DailyUsageResponse(BaseModel):
    """Daily usage breakdown."""

    date: date
    total_prompt_tokens: int
    total_completion_tokens: int
    total_cached_tokens: int
    total_cost: float
    llm_call_count: int
    tool_call_count: int
    tokens_by_model: dict[str, int]
    cost_by_model: dict[str, float]


class DailyUsageListResponse(BaseModel):
    """List of daily usage records."""

    days: list[DailyUsageResponse]
    total_days: int


class MonthlyUsageDetailResponse(BaseModel):
    """Monthly usage detail."""

    year: int
    month: int
    total_prompt_tokens: int
    total_completion_tokens: int
    total_cached_tokens: int
    total_cost: float
    total_llm_calls: int
    total_tool_calls: int
    total_sessions: int
    active_days: int
    cost_by_model: dict[str, float]


class MonthlyUsageListResponse(BaseModel):
    """List of monthly usage records."""

    months: list[MonthlyUsageDetailResponse]
    total_months: int


class SessionUsageResponse(BaseModel):
    """Usage for a specific session."""

    session_id: str
    total_prompt_tokens: int
    total_completion_tokens: int
    total_cached_tokens: int
    total_cost: float
    llm_call_count: int
    tool_call_count: int
    tokens_by_model: dict[str, int]
    cost_by_model: dict[str, float]
    first_activity: datetime | None
    last_activity: datetime | None


class ModelPricingResponse(BaseModel):
    """Pricing for a single model."""

    model: str
    prompt_price: float  # Per 1M tokens
    completion_price: float  # Per 1M tokens
    cached_price: float | None  # Per 1M tokens (if available)


class PricingListResponse(BaseModel):
    """List of all model pricing."""

    models: list[ModelPricingResponse]


class AgentUsageSummaryResponse(BaseModel):
    """Run-aware usage summary for the selected period."""

    run_count: int
    completed_run_count: int
    failed_run_count: int
    success_rate: float
    avg_run_duration_ms: float
    total_cost: float
    total_input_tokens: int
    total_cached_input_tokens: int
    total_output_tokens: int
    total_reasoning_tokens: int
    total_tool_calls: int
    total_mcp_calls: int
    cache_savings_estimate: float


class AgentRunResponse(BaseModel):
    """Recent agent run row for the usage table."""

    run_id: str
    session_id: str
    started_at: datetime
    completed_at: datetime | None
    status: str
    duration_ms: float | None
    total_cost: float
    total_tokens: int
    tool_call_count: int
    mcp_call_count: int
    primary_model: str | None
    primary_provider: str | None


class AgentRunListResponse(BaseModel):
    """List wrapper for recent agent runs."""

    runs: list[AgentRunResponse]
    total_runs: int


class AgentUsageBreakdownRowResponse(BaseModel):
    """Breakdown row grouped by model/provider/tool/MCP server."""

    key: str
    run_count: int
    input_tokens: int
    cached_input_tokens: int
    output_tokens: int
    reasoning_tokens: int
    cost: float
    avg_duration_ms: float
    error_rate: float


class AgentUsageBreakdownListResponse(BaseModel):
    """List wrapper for grouped agent usage breakdowns."""

    rows: list[AgentUsageBreakdownRowResponse]
    total_rows: int


class AgentUsageTimeseriesPointResponse(BaseModel):
    """Daily trend point for the usage chart."""

    date: datetime
    run_count: int
    success_count: int
    failed_count: int
    cost: float
    input_tokens: int
    cached_input_tokens: int
    output_tokens: int
    reasoning_tokens: int
    tool_calls: int
    mcp_calls: int


class AgentUsageTimeseriesResponse(BaseModel):
    """List wrapper for agent usage trend points."""

    points: list[AgentUsageTimeseriesPointResponse]
    total_points: int

File: frontend/src/api/usage.ts
import { apiClient } from './client'

interface RequestOptions {
  signal?: AbortSignal
}

// Today's usage summary
export interface TodayUsage {
  tokens: number
  cost: number
  llm_calls: number
  tool_calls: number
}

// This month's usage summary
export interface MonthUsage {
  tokens: number
  cost: number
  llm_calls: number
  tool_calls: number
  sessions: number
  active_days: number
}

// Combined usage summary
export interface UsageSummary {
  today: TodayUsage
  month: MonthUsage
}

// Daily usage record
export interface DailyUsage {
  date: string
  total_prompt_tokens: number
  total_completion_tokens: number
  total_cached_tokens: number
  total_cost: number
  llm_call_count: number
  tool_call_count: number
  tokens_by_model: Record<string, number>
  cost_by_model: Record<string, number>
}

// Daily usage list response
export interface DailyUsageList {
  days: DailyUsage[]
  total_days: number
}

// Monthly usage detail
export interface MonthlyUsageDetail {
  year: number
  month: number
  total_prompt_tokens: number
  total_completion_tokens: number
  total_cached_tokens: number
  total_cost: number
  total_llm_calls: number
  total_tool_calls: number
  total_sessions: number
  active_days: number
  cost_by_model: Record<string, number>
}

// Monthly usage list response
export interface MonthlyUsageList {
  months: MonthlyUsageDetail[]
  total_months: number
}

// Session usage
export interface SessionUsage {
  session_id: string
  total_prompt_tokens: number
  total_completion_tokens: number
  total_cached_tokens: number
  total_cost: number
  llm_call_count: number
  tool_call_count: number
  tokens_by_model: Record<string, number>
  cost_by_model: Record<string, number>
  first_activity: string | null
  last_activity: string | null
}

// Model pricing info
export interface ModelPricing {
  model: string
  prompt_price: number
  completion_price: number
  cached_price: number | null
}

// Pricing list response
export interface PricingList {
  models: ModelPricing[]
}

export type AgentUsageRange = '7d' | '30d' | '90d'
export type AgentUsageBreakdownGroup = 'model' | 'provider' | 'tool' | 'mcp_server'

export interface AgentUsageSummary {
  run_count: number
  completed_run_count: number
  failed_run_count: number
  success_rate: number
  avg_run_duration_ms: number
  total_cost: number
  total_input_tokens: number
  total_cached_input_tokens: number
  total_output_tokens: number
  total_reasoning_tokens: number
  total_tool_calls: number
  total_mcp_calls: number
  cache_savings_estimate: number
}

export interface AgentUsageRun {
  run_id: string
  session_id: string
  started_at: string
  completed_at: string | null
  status: string
  duration_ms: number | null
  total_cost: number
  total_tokens: number
  tool_call_count: number
  mcp_call_count: number
  primary_model: string | null
  primary_provider: string | null
}

export interface AgentUsageRunList {
  runs: AgentUsageRun[]
  total_runs: number
}

export interface AgentUsageBreakdownRow {
  key: string
  run_count: number
  input_tokens: number
  cached_input_tokens: number
  output_tokens: number
  reasoning_tokens: number
  cost: number
  avg_duration_ms: number
  error_rate: number
}

export interface AgentUsageBreakdownList {
  rows: AgentUsageBreakdownRow[]
  total_rows: number
}

export interface AgentUsageTimeseriesPoint {
  date: string
  run_count: number
  success_count: number
  failed_count: number
  cost: number
  input_tokens: number
  cached_input_tokens: number
  output_tokens: number
  reasoning_tokens: number
  tool_calls: number
  mcp_calls: number
}

export interface AgentUsageTimeseries {
  points: AgentUsageTimeseriesPoint[]
  total_points: number
}

/**
 * Get usage summary for today and this month
 */
export async function getUsageSummary(options: RequestOptions = {}): Promise<UsageSummary> {
  const response = await apiClient.get<{ data: UsageSummary }>('/usage/summary', options)
  return response.data.data
}

/**
 * Get daily usage breakdown
 */
export async function getDailyUsage(days: number = 30, options: RequestOptions = {}): Promise<DailyUsageList> {
  const response = await apiClient.get<{ data: DailyUsageList }>(`/usage/daily?days=${days}`, options)
  return response.data.data
}

/**
 * Get monthly usage summaries
 */
export async function getMonthlyUsage(months: number = 12): Promise<MonthlyUsageList> {
  const response = await apiClient.get<{ data: MonthlyUsageList }>(`/usage/monthly?months=${months}`)
  return response.data.data
}

/**
 * Get usage for a specific session
 */
export async function getSessionUsage(sessionId: string): Promise<SessionUsage> {
  const response = await apiClient.get<{ data: SessionUsage }>(`/usage/session/${sessionId}`)
  return response.data.data
}

/**
 * Get model pricing information
 */
export async function getModelPricing(): Promise<PricingList> {
  const response = await apiClient.get<{ data: PricingList }>('/usage/pricing')
  return response.data.data
}

/**
 * Get run-aware usage summary for the selected time range
 */
export async function getAgentUsageSummary(
  range: AgentUsageRange = '30d',
  options: RequestOptions = {}
): Promise<AgentUsageSummary> {
  const response = await apiClient.get<{ data: AgentUsageSummary }>(
    `/usage/agent/summary?range=${range}`,
    options
  )
  return response.data.data
}

/**
 * Get recent agent runs
 */
export async function getAgentUsageRuns(
  range: AgentUsageRange = '30d',
  limit: number = 20,
  options: RequestOptions = {}
): Promise<AgentUsageRunList> {
  const response = await apiClient.get<{ data: AgentUsageRunList }>(
    `/usage/agent/runs?range=${range}&limit=${limit}`,
    options
  )
  return response.data.data
}

/**
 * Get grouped agent usage breakdown
 */
export async function getAgentUsageBreakdown(
  range: AgentUsageRange = '30d',
  groupBy: AgentUsageBreakdownGroup = 'model',
  options: RequestOptions = {}
): Promise<AgentUsageBreakdownList> {
  const response = await apiClient.get<{ data: AgentUsageBreakdownList }>(
    `/usage/agent/breakdown?range=${range}&group_by=${groupBy}`,
    options
  )
  return response.data.data
}

/**
 * Get run-aware daily usage trends
 */
export async function getAgentUsageTimeseries(
  range: AgentUsageRange = '30d',
  bucket: 'hour' | 'day' | 'week' = 'day',
  options: RequestOptions = {}
): Promise<AgentUsageTimeseries> {
  const response = await apiClient.get<{ data: AgentUsageTimeseries }>(
    `/usage/agent/timeseries?range=${range}&bucket=${bucket}`,
    options
  )
  return response.data.data
}
